File: MetalSampleCodeLibrary/RenderWorkflows/ImplementingOrderIndependentTransparencyWithImageBlocks/Application/AAPLViewController.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
The app's main view controller.
*/

#import "AAPLViewController.h"
#import "AAPLRenderer.h"

@implementation AAPLViewController
{
    MTKView* _mtkView;
    AAPLRenderer* _renderer;

#if TARGET_OS_IPHONE

    __weak IBOutlet UISegmentedControl *_orderIndependentTransparencyControl;

#elif TARGET_OS_OSX

    __weak IBOutlet NSSegmentedControl *_orderIndependentTransparencyControl;

#endif
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    _mtkView = (MTKView*)self.view;
    _mtkView.device = MTLCreateSystemDefaultDevice();

    NSAssert(_mtkView.device, @"Metal is not supported on this device");

    _renderer = [[AAPLRenderer alloc] initWithMetalKitView:_mtkView];

    NSAssert(_renderer, @"Renderer failed to initialize");

    [_renderer mtkView:_mtkView drawableSizeWillChange:_mtkView.drawableSize];

    _mtkView.delegate = _renderer;

    if (_renderer.supportsOrderIndependentTransparency)
    {
        _orderIndependentTransparencyControl.enabled = YES;
    }
    else
    {
        _orderIndependentTransparencyControl.enabled = NO;
    }

    NSUInteger segmentSelectionIndex = (_renderer.enableOrderIndependentTransparency) ? 0 : 1;

#if TARGET_OS_IPHONE
    _orderIndependentTransparencyControl.selectedSegmentIndex = segmentSelectionIndex;
#else
    [_orderIndependentTransparencyControl setSelected:YES forSegment:segmentSelectionIndex];
#endif
}

- (BOOL)acceptsFirstResponder
{
    return YES;
}

#if TARGET_OS_IPHONE

- (IBAction)updateTransparencyMethod:(UISegmentedControl *)transparencyControl
{
    if (!transparencyControl.selectedSegmentIndex)
    {
        _renderer.enableOrderIndependentTransparency = YES;
    }
    else
    {
        _renderer.enableOrderIndependentTransparency = NO;
    }
}

- (IBAction)toggleRotation:(UISwitch *)rotationControl
{
    if (rotationControl.on)
    {
        _renderer.enableRotation = YES;
    }
    else
    {
        _renderer.enableRotation = NO;
    }
}

#elif TARGET_OS_OSX

- (IBAction)updateTransparencyMethod:(NSSegmentedControl *)transparencyControl
{
    if (!transparencyControl.selectedSegment)
    {
        _renderer.enableOrderIndependentTransparency = YES;
    }
    else
    {
        _renderer.enableOrderIndependentTransparency = NO;
    }
}

- (IBAction)toggleRotation:(NSSwitch *)rotationControl
{
    if (rotationControl.state == NSControlStateValueOn)
    {
        _renderer.enableRotation = YES;
    }
    else
    {
        _renderer.enableRotation = NO;
    }
}

- (void)viewDidAppear
{
    // Set the view to handle Key events by making it the window's first responder.
    [_mtkView.window makeFirstResponder:self];
}

#endif

@end

File: MetalFX/ApplyingTemporalAntialiasingAndUpscalingUsingMetalFX/Application/macOS/AAPLViewController.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
This file provides the view controller for the MetalFX sample.
*/

import Cocoa
import MetalKit

/// The macOS-specific view controller.
class AAPLViewController: NSViewController {
    var renderer: AAPLRenderer!
    var mtkView: MTKView!

    @IBOutlet var animationToggleButton: NSButton!
    @IBOutlet var resetHistoryToggleButton: NSButton!
    @IBOutlet var proceduralTextureToggleButton: NSButton!
    @IBOutlet var scalingModeComboButton: NSComboButton!
    @IBOutlet var scalingModeSegmentedControl: NSSegmentedControl!
    @IBOutlet var mipBiasControlsLabel: NSTextField!
    @IBOutlet var mipBiasLabel: NSTextField!
    @IBOutlet var mipBiasSlider: NSSlider!
    @IBOutlet var renderScaleLabel: NSTextField!
    @IBOutlet var renderScaleSlider: NSSlider!

    override func viewDidLoad() {
        super.viewDidLoad()

        guard let mtkView = view as? MTKView else {
            print("The view attached to AAPLViewController is not an MTKView.")
            return
        }

        // Select the default Metal device.
        guard let defaultDevice = MTLCreateSystemDefaultDevice() else {
            print("Metal is not supported on this device.")
            return
        }

        mtkView.device = defaultDevice

        guard let newRenderer = AAPLRenderer(metalKitView: mtkView) else {
            print("The app is unable to create the renderer.")
            return
        }

        renderer = newRenderer

        renderer.mtkView(mtkView, drawableSizeWillChange: mtkView.drawableSize)

        mtkView.delegate = renderer

        optionsChanged(nil)
        scalingModeChanged(nil)
    }

    @IBAction func optionsChanged(_ sender: Any?) {
        renderer.animationEnabled = (animationToggleButton.state == .on)
        renderer.resetHistoryEnabled = (resetHistoryToggleButton.state == .on)
        renderer.proceduralTextureEnabled = (proceduralTextureToggleButton.state == .on)
        renderer.adjustRenderScale(renderScaleSlider.floatValue)
        renderScaleSlider.floatValue = renderer.renderTarget.renderScale
        renderScaleLabel.stringValue = String(format: "% 3d%%", arguments: [Int(renderScaleSlider.floatValue * 100)])
        
        if renderer.proceduralTextureEnabled {
            renderer.textureMipmapBias = 0
            mipBiasControlsLabel.isHidden = true
            mipBiasLabel.isHidden = true
            mipBiasSlider.isHidden = true
        } else {
            mipBiasControlsLabel.isHidden = false
            mipBiasLabel.isHidden = false
            mipBiasSlider.isHidden = false
            renderer.textureMipmapBias = mipBiasSlider.floatValue
            mipBiasLabel.stringValue = String(format: "% 1.3f", arguments: [mipBiasSlider.floatValue])
        }
    }

    @IBAction func scalingModeChanged(_ sender: Any?) {
        // Save the current mode to configure MetalFX if the combo button changes.
        let previousScalingMode = renderer.mfxScalingMode
        
        let index = scalingModeSegmentedControl.selectedSegment
        let currentScalingMode = AAPLScalingMode(rawValue: index)!
        renderer.mfxScalingMode = currentScalingMode
        
        if previousScalingMode != renderer.mfxScalingMode {
            renderer.setupMetalFX()
        }
        
        // If the effect isn’t available, the renderer picks a supported effect.
        // Set the control to the current applied effect.
        if renderer.mfxScalingMode != currentScalingMode {
            scalingModeSegmentedControl.selectedSegment = renderer.mfxScalingMode.rawValue
        }
    }
}
